packages/datetime/src/datePicker.tsx

```diff
@@ -221,153 +186,146 @@ export class DatePicker extends AbstractPureComponent<IDatePickerProps, IDatePic
     private renderOptionsBar() {
         return (
             <div className={Classes.DATEPICKER_FOOTER}>
-                <Button className={CoreClasses.MINIMAL} onClick={this.handleTodayClick} text="Today" />
-                <Button className={CoreClasses.MINIMAL} onClick={this.handleClearClick} text="Clear" />
+                <Button minimal={true} onClick={this.handleTodayClick} text="Today" />
+                <Button minimal={true} onClick={this.handleClearClick} text="Clear" />
             </div>
         );
     }
 
     private handleDayClick = (day: Date, modifiers: DayModifiers, e: React.MouseEvent<HTMLDivElement>) => {
         Utils.safeInvoke(this.props.dayPickerProps.onDayClick, day, modifiers, e);
 
-        let newValue = day;
+        // allow toggling selected date by clicking it again (if prop enabled)
+        const newValue = this.props.canClearSelection && modifiers.selected ? null : day;
 
-        if (this.props.canClearSelection && modifiers.selected) {
-            newValue = null;
-        }
-
-        if (this.props.value === undefined) {
-            // component is uncontrolled
-            if (!modifiers.disabled) {
-                const displayMonth = day.getMonth();
-                const displayYear = day.getFullYear();
-                const selectedDay = day.getDate();
+        if (!modifiers.disabled) {
+            if (this.props.value === undefined) {
+                // set now if uncontrolled, otherwise they'll be updated in `componentWillReceiveProps`
                 this.setState({
-                    displayMonth,
-                    displayYear,
-                    selectedDay,
-                    value: newValue,
+                    displayMonth: day.getMonth(),
+                    displayYear: day.getFullYear(),
+                    selectedDay: day.getDate(),
                 });
             }
-        }
-
-        if (!modifiers.disabled) {
-            Utils.safeInvoke(this.props.onChange, newValue, true);
-            if (this.state.value != null && this.state.value.getMonth() !== day.getMonth()) {
+            if (this.state.value == null || this.state.value.getMonth() !== day.getMonth()) {
                 this.ignoreNextMonthChange = true;
             }
-        } else {
-            // rerender base component to get around bug where you can navigate past bounds by clicking days
-            this.forceUpdate();
+            this.updateValue(newValue, true);
         }
     };
 
     private computeValidDateInSpecifiedMonthYear(displayYear: number, displayMonth: number): Date {
         const { minDate, maxDate } = this.props;
+        // month is 0-based, date is 1-based. date 0 is last day of previous month.
```

Comment:
RE: `date 0 is last day of previous month`: how so? is that actually how the API works?

Reply:
sure seems to be!